pkg/alert/rule_controller.go

```diff
@@ -0,0 +1,465 @@
+		return nil
+
+	case err != nil:
+		return err
+	}
+
+	klog.V(4).Infof("Syncing AlertingRule: %s", key)
+
+	rule.APIVersion = osmv1alpha1.GroupVersion.String()
+	rule.Kind = "AlertingRule"
+
+	// Generate the new or updated PrometheusRule object.
+	promRule := &monv1.PrometheusRule{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      rc.promRuleName(namespace, name),
+			Namespace: rc.client.Namespace(),
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion:         rule.APIVersion,
+					Kind:               rule.Kind,
+					Name:               rule.Name,
+					UID:                rule.UID,
+					Controller:         pointer.Bool(true),
+					BlockOwnerDeletion: pointer.Bool(true),
+				},
+			},
+			Labels: map[string]string{
+				"app.kubernetes.io/version":   rc.version,
+				"app.kubernetes.io/component": "alerting-rules-controller",
+				"app.kubernetes.io/name":      "cluster-monitoring-operator",
+				"app.kubernetes.io/part-of":   "openshift-monitoring",
+				"prometheus":                  "k8s",
+				"role":                        "alerting-rules",
+			},
+		},
+		Spec: monv1.PrometheusRuleSpec{
+			Groups: rc.convertRuleGroups(rule.Spec.Groups),
+		},
+	}
+
+	if err := rc.client.CreateOrUpdatePrometheusRule(ctx, promRule); err != nil {
+		return err
+	}
+
+	// Update AlertingRule status with latest observed generation, and the name of
+	// the generated PrometheusRule object.
+	rule.Status.ObservedGeneration = rule.Generation
+	rule.Status.PrometheusRule = fmt.Sprintf("%s/%s", promRule.Namespace, promRule.Name)
+
+	return rc.client.UpdateAlertingRuleStatus(ctx, rule)
+}
+
+// promRuleName returns the name of PrometheusRule to be generated for the given
+// AlertingRule namespace and name.
+func (rc *RuleController) promRuleName(namespace, name string) string {
+	return fmt.Sprintf("alerts-%s-%s", namespace, name)
+}
+
+// convertRuleGroups converts the given OpenShift Monitoring RuleGroups to their
+// corresponding upstream prometheus-operator versions.  It ensures each rule has
+// a static label identifying the rule as coming from platform-monitoring, while
+// being user-defined.
+func (rc *RuleController) convertRuleGroups(groups []osmv1alpha1.RuleGroup) []monv1.RuleGroup {
+	monv1Groups := make([]monv1.RuleGroup, len(groups))
+
+	for i, group := range groups {
+		monv1Group := monv1.RuleGroup{Name: group.Name, Interval: group.Interval}
+		monv1Group.Rules = make([]monv1.Rule, len(group.Rules))
+
+		for j, rule := range group.Rules {
+			monv1Rule := monv1.Rule{
+				Alert:       rule.Alert,
+				Expr:        rule.Expr,
+				For:         rule.For,
+				Labels:      rule.Labels,
+				Annotations: rule.Annotations,
+			}
+
+			if monv1Rule.Labels == nil {
+				monv1Rule.Labels = make(map[string]string)
+			}
+
+			// Set a static label indicating this rule comes from
+			// platform-monitoring, but is user defined.
+			monv1Rule.Labels[alertSourceLabel] = alertSourceValue
```

Comment:
I realized this would be overwritten by the default relabel rules anyway. I've changed it to `openshift_io_user_alert=true` for now, but of course I'm open to suggestions.